improver/utilities/ancillary_creation.py

```diff
@@ -158,87 +192,107 @@ def unnormalised_smoothing_coefficients(self, gradient_cube):
-                    orography-dependent smoothing_coefficients calculated in
-                    the y direction.
+            smoothing_coefficient_x (iris.cube.Cube):
+                Smoothing coefficients calculated along the x-dimension.
+            smoothing_coefficient_y (iris.cube.Cube):
+                Smoothing coefficients calculated along the y-dimension.
+            mask (iris.cube.Cube):
+                The mask defining areas in which smoothing coefficients should
+                be zeroed.
         """
-        smoothing_coefficient_x = self.unnormalised_smoothing_coefficients(gradient_x)
-        smoothing_coefficient_y = self.unnormalised_smoothing_coefficients(gradient_y)
-
-        (
-            smoothing_coefficient_x,
-            smoothing_coefficient_y,
-        ) = self.scale_smoothing_coefficients(
-            [smoothing_coefficient_x, smoothing_coefficient_y],
-            min_output=self.min_smoothing_coefficient,
-            max_output=self.max_smoothing_coefficient,
-        )
-        smoothing_coefficient_x = self.update_smoothing_coefficients_metadata(
-            smoothing_coefficient_x, "smoothing_coefficient_x"
-        )
-        smoothing_coefficient_y = self.update_smoothing_coefficients_metadata(
-            smoothing_coefficient_y, "smoothing_coefficient_y"
-        )
-
-        return smoothing_coefficient_x, smoothing_coefficient_y
+        if self.use_mask_boundary:
+            zero_points_x = np.diff(mask.data, axis=1) != 0
+            zero_points_y = np.diff(mask.data, axis=0) != 0
+        else:
+            zero_points_x = self.mask_comparison(
+                mask.data[:, :-1] + mask.data[:, 1:], 1
+            )
+            zero_points_y = self.mask_comparison(
+                mask.data[:-1, :] + mask.data[1:, :], 1
+            )
+        smoothing_coefficient_x.data[zero_points_x] = 0.0
+        smoothing_coefficient_y.data[zero_points_y] = 0.0
 
-    def process(self, cube):
+    def process(self, cube, mask=None):
         """
         This creates the smoothing_coefficient cubes. It returns one for the x
         direction and one for the y direction. It uses the
         DifferenceBetweenAdjacentGridSquares plugin to calculate an average
-        gradient across each grid square.  These gradients are then used to
+        gradient across each grid square. These gradients are then used to
         calculate "smoothing_coefficient" arrays that are normalised between a
         user-specified max and min.
 
         Args:
             cube (iris.cube.Cube):
                 A 2D field of orography on the grid for which
                 smoothing_coefficients are to be generated.
+            mask (iris.cube.Cube or None):
+                A mask that defines where the smoothing coefficients should
+                be zeroed. How the mask is used is determined by the plugin
+                configuration arguments.
```

Comment:
It might be worth adding here that the mask is expected to have the same spatial dimensions are cube, given that an exception is raised for this below.